src/main/java/no/digipost/io/InputStreamIterator.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) Posten Norge AS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package no.digipost.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * InputStreamIterator is an {@link Iterator} reading from an {@link InputStream} in chunks
+ * where each chunk is returned as the next element in the iterable.
+ * When the input stream is fully consumed the iterator has no more elements.
+ */
+public class InputStreamIterator implements Iterator<byte[]>, AutoCloseable {
+    private final InputStream inputStream;
+    private final int chunkSize;
+    private byte[] next;
+    private Boolean hasNext;
+    private boolean endOfStreamReached = false;
+
+    public InputStreamIterator(InputStream inputStream, int chunkSize) {
+        this.inputStream = inputStream;
+        this.chunkSize = chunkSize;
+    }
+
+    private byte[] loadNextChunk() {
+        if (endOfStreamReached) return null;
+
+        byte[] chunk = new byte[chunkSize];
+        int bytesRead = 0;
+        try {
+            bytesRead = inputStream.read(chunk);
+            if (bytesRead == -1) {
+                endOfStreamReached = true;
+                inputStream.close();
+                return null;
+            }
+        } catch (IOException e) {
+            throw new WrappedInputStreamFailed(e, inputStream);
+        }
+
+        if (bytesRead < chunkSize) {
+            // resize the buffer if less data was read
+            byte[] smallerBuffer = new byte[bytesRead];
+            System.arraycopy(chunk, 0, smallerBuffer, 0, bytesRead);
+            chunk = smallerBuffer;
+        }
+
+        return chunk;
+    }
+
+    /**
+     *  If the iterator fails reading from the wrapped InputStream an 
+     *  {@link  InputStreamIterator.WrappedInputStreamFailed} runtime exception is thrown.
+     *  
+     * @return true if the iteration has more elements
+     */
+    @Override
+    public boolean hasNext() {
+        if (hasNext == null) {
+            next = loadNextChunk();
+            hasNext = (next != null);
+        }
+
+        return hasNext;
```

Comment:
This juggling with a "tri-state boolean" (true, false, null) really tickles. (I already struggle with keeping up with two-state booleans, a.k.a. `boolean`). Had to do some effort to assign semantics to each state:
- `null`: next chunk may be attempted loaded, and this determines if next chunk is actually available
- `true`: next chunk is loaded, and ready to be supplied with `.next()`
- `false`: next chunk has been attempted loaded, but no more chunks was available, so there are no more chunks left.

I all fine with this, as long as it is encapsulated. Just needed some mental exercise to wrap my head around it. More exposure to mental exercises isn't necessarily a bad thing 😛

Reply:
I guess George Boole would turn in his grave if he saw that his boolean value it self can be undefined. I guess hasNext could be expressed with primitives if you use two booleans with some boolean algebra. But i am not convinced that it will be more readable. :D